### DISCUSSION TITLE
is it possible to execute a script pre-compile time?

### DISCUSSION THREAD

**Discussion by michelrandahl** [purs-nix/purs-nix#51] [Q&A] [answered]
I am in the middle of writing my very first purescript application and have been using `nix develop` all the time while working on it, however I realized that it would be nice to be able to properly build and install the application through nix. The problem though, is that I have decided to write some FFI code in typescript, and my workflow during development has simply been to invoke `tsc` to generate the js files before building or running my code. My knowledge on flake and Nix doesn't seem sufficient for figuring out how to accomplish this in my flake.nix file.
Is it possible to somehow execute `tsc` before compilation when running for example `nix build`.
If this is something that is considered a trivial Nix thing, then I would much appreciate links to tutorials, books or other learning material.

**Comment by ursi** [comment, marked as answer]
Yes, you definitely should be able to do this. it would likely look something like this
```nix
pkgs.runCommand "js" {} ''
  # lets say all your js files are in a directory called `js-files`  
  cp -r ${./js-files}/. . # copy them into our derivation environment
  ${lib.getExe pkgs.tsc} # run typescript
  mkdir $out # create our output directory
  mv ... # move generated js files into `$out`
''
```
If you're just making a single js file, then you don't need to move it into `$out`, you can just name that file `$out` and your derivation will resolve to be that file, instead of a directory of files.

You can use this derivation with purs-nix to link it into your package directly, with this api https://github.com/purs-nix/purs-nix/blob/master/docs/foreign.md
